homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,263 @@
+    MachineState.FactoryDiagnostic: "Factory Diagnostic Mode",
+    MachineState.LifeTest: "Life Test",
+    MachineState.CustomerFocusMode: "Customer Focus Mode",
+    MachineState.DemoMode: "Demo Mode",
+    MachineState.HardStopOrError: "Hard Stop or Error",
+    MachineState.SystemInit: "System Initialize",
+}
+
+CYCLE_FUNC = [
+    (WasherDryer.get_cycle_status_filling, "Cycle Filling"),
+    (WasherDryer.get_cycle_status_rinsing, "Cycle Rinsing"),
+    (WasherDryer.get_cycle_status_sensing, "Cycle Sensing"),
+    (WasherDryer.get_cycle_status_soaking, "Cycle Soaking"),
+    (WasherDryer.get_cycle_status_spinning, "Cycle Spinning"),
+    (WasherDryer.get_cycle_status_washing, "Cycle Washing"),
+]
+
+
+ICON_D = "mdi:tumble-dryer"
+ICON_W = "mdi:washing-machine"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryerClass) -> str | None:
+    """Determine correct states for a washer."""
+
+    if washer.washdry.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        return "Door open"
+
+    machine_state = washer.washdry.get_machine_state()
+    machine_cycle = None
+
+    for func, cycle_name in CYCLE_FUNC:
+        if func(washer.washdry):
+            machine_cycle = cycle_name
+            break
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+def completion_time(washer: WasherDryerClass) -> datetime | None:
+    """Calculate the time stamp for completion."""
+    machine_state = washer.washdry.get_machine_state()
+    if (
+        machine_state.value in {MachineState.Complete.value, MachineState.Standby.value}
+        and washer.running
+    ):
+        washer.running = False
+        washer.timestamp = utcnow()
+
+    if machine_state is MachineState.RunningMainCycle:
+        washer.running = True
+        washer.timestamp = utcnow() + timedelta(
+            seconds=int(
+                washer.washdry.get_attribute("Cavity_TimeStatusEstTimeRemaining")
+            )
+        )
+    return washer.timestamp
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="State",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="timeremaining",
+        name="End Time",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement="s",
```

Comment:
I'm not sure this is correct to provide anymore

Reply:
I'm not sure I understand your reference here (sorry).

Reply:
Oh -duh unit_of_measurment DNA for TIMESTAMP.